File: 0056-merge-intervals/0056-merge-intervals.cpp
class Solution
{
    public:
        vector<vector < int>> merge(vector<vector < int>> &intervals)
        {
            vector<vector < int>> ans;
            int n = intervals.size();
            sort(intervals.begin(), intervals.end());
            vector<int> curr = intervals[0];
            for (int i = 1; i < n; i++)
            {
               	//no overlap
                if (curr[1] < intervals[i][0])
                {
                    ans.push_back(curr);
                    curr = intervals[i];
                }
                else
                {
                    curr[1] = max(curr[1], intervals[i][1]);
                }
            }
            ans.push_back(curr);
            return ans;
        }
};

//[[1,13], [2,5], [6,9], [10,12], [12,15], [18,22], [23,25]]

File: 0056-merge-intervals/0056-merge-intervals.java

public class Solution {
    public int[][] merge(int[][] intervals) {
        List<int[]> ans = new ArrayList<>();
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        int s = intervals[0][0];
        int e = intervals[0][1];

        for (int i = 0; i < intervals.length; i++) {
            int ns = intervals[i][0];
            int ne = intervals[i][1];

            // in range of previous interval
            if (ns <= e) {
                e = Math.max(e, ne);
            } else {
                ans.add(new int[]{s, e});
                s = ns;
                e = ne;
            }
        }
        ans.add(new int[]{s, e});


        int[][] result = new int[ans.size()][2];
        for (int i = 0; i < ans.size(); i++) {
            result[i] = ans.get(i);
        }

        return result;
    
    }
}
